%\VignetteIndexEntry{siggenes Manual}
%\VignetteKeywords{Expression Analysis}
%\VignetteDepends{siggenes}
%\VignettePackage{siggenes}


\documentclass[a4paper,12pt]{article}
\usepackage{amsmath,amssymb,pstricks}
\usepackage{graphicx}


%\setlength{\parindent}{0cm} \setlength{\parskip}{18pt}

\renewcommand{\baselinestretch}{1.2}


\begin{document}
\title{Identifying differentially expressed genes with
\texttt{siggenes}}

\author{Holger Schwender\\  holger.schw@gmx.de}
\date{}

\maketitle

%\setlength{\parskip}{0pt}

\vspace{12pt}

\begin{abstract}
\noindent In this vignette, we show how the functions contained in
the \texttt{R} package \texttt{siggenes} can be used to perform
both the Significance Analysis of Microarrays (SAM) proposed by
Tusher et al.\ (2001) and the Empirical Bayes Analysis of
Microarrays (EBAM) suggested by Efron et al.\ (2001).
\end{abstract}
\thispagestyle{empty}

%\setlength{\parskip}{18pt}

\vspace*{48pt} \centerline{\large \bf{PLEASE NOTE}} \vspace*{18pt}
\noindent Since there is a patent pending for the Significance
Analysis of Microarrays (SAM), this package is only free for
non-commercial users. Non-academic users \textbf{MUST} have a
valid license for the full (Excel) version of the SAM software
programmed at Stanford University, see
\texttt{http://www-stat.stanford. edu/$\sim$tibs/SAM/index.html}.


\newpage

\section{Introduction}

Both the Significance Analysis of Microarrays (SAM) proposed by
Tusher et al.\ (2001) and the Empirical Bayes Analysis of
Microarrays (EBAM) suggested by Efron et al.\ (2001) can be used
to identify differentially expressed genes and to estimate the
False Discovery Rate (FDR). The \texttt{R} package
\texttt{siggenes} contains functions for both a SAM and an EBAM
analysis using either a modified $t$ statistic or the Wilcoxon rank
sum statistics. In this vignette, it is described how these
functions can be used. For details on the algorithms
behind these functions, see Schwender (2003).

As usual, it is necessary to load the package.

<<echo=F,results=hide>>=
library(siggenes)
@

<<>>=
library(siggenes)
@



In the following, we use the Golub et al.\ (1999) data set as it
is provided by the \texttt{multtest} package to illustrate how the
SAM and the EBAM analyses can be performed.

<<>>=
library(multtest)
data(golub)
@

\texttt{data(golub)} consists of a 3051x38 matrix \texttt{golub}
containing the expression levels of 3051 genes and 38 samples, a
vector \texttt{golub.cl} containing the class labels of the 38
samples, and a 3051x3 matrix \texttt{golub.gnames} whose second
column consists of the names of the genes.

\section{Required Arguments: \texttt{data} and \texttt{cl}}

In the first step of each of the four analyses (SAM and EBAM using
the modified $t$ statistic, and SAM and EBAM using the Wilcoxon
rank sum statistic), two arguments are required: \texttt{data} and
\texttt{cl}.

The first required argument, \texttt{data}, is the matrix (or the
data frame) containing the gene expression data that should be
analyzed. Each row of this matrix must correspond to a gene, and
each column must correspond to a sample.

The second required argument, \texttt{cl}, is the vector
containing the class labels of the samples. The correct
specification of the class labels depends on the type of data that
should be analyzed. On the basis of this specification, the
functions identify the type of data automatically.

\vspace*{18pt} \noindent \textbf{Two class, unpaired data.} In
this case, the functions expect a vector \texttt{cl} consisting
only of 0's and 1's, where all the samples with class label '0'
belong to one group (e.g., the control group), and the samples
with class label '1' belong to the other group (e.g., the
case group). So, e.g., if the first \texttt{n1}$=5$ columns of the
data matrix correspond to controls and the next \texttt{n2}$=5$
columns correspond to cases, then the vector \texttt{cl} is given
by

<<>>=
n1 <- n2 <- 5
rep(c(0, 1), c(n1, n2))
@

The functions also accept other values than 0 and 1. In this case,
the smaller value is automatically set to 0, and the larger value
to 1. So if, e.g., 1 is used as the label for group 1, and 2 for
the label of group 2, then the functions will automatically set the
class label '1' to 0, and the class label '2' to 1.


\vspace*{18pt} \noindent \textbf{Two class, paired data.} Denoting
the number of columns/samples by $n$, we here have $K=n/2$ paired
observations. Each of the $K$ samples belonging to the first group
(e.g., the after treatment group) is labelled by one of the
integers between 1 and $K$, and each of the $K$ samples belonging
to the other group (e.g., the before treatment group) is labelled
by one of the integers between -1 and $-K$, where the sample with
class label '$k$' and the sample with label $'-k'$ build an
observation pair, $k=1,\ldots,K$. So if, e.g., the first $k=5$
columns of the data matrix contain samples from the before
treatment group, and the next $k=5$ columns contain samples from
the after treatment group, where the samples 1 and 6, 2 and 7,
..., respectively, build a pair, then the vector \texttt{cl} is
given by

<<>>=
k <- 5
c((-1:-5), 1:5)
@

Another example: If the first column contains the before treatment
measurements of an observation, the second column the after
treatment measurements of the same observation, the third column
the before treatment measurements of the second observations, the
fourth column the after treatment measurements of the second
observation, and so on, then a possible way to generate the vector
\texttt{cl} for $k=5$ paired observations is

<<>>=
k <- 5
rep(1:k, e = 2) * rep(c(-1, 1), k)
@


\vspace*{18pt}
\noindent \textbf{One class data.} It is also
possible to perform a SAM analysis (using the modified $t$
statistic) for one class data. In this case, the function
\texttt{sam} expects a vector of length $n$ containing
only 1's, where $n$ denotes the number of samples/columns, but
\texttt{sam} also accepts another value. In the latter case,
\texttt{sam} automatically sets this value to 1. So for $n=10$,
the vector \texttt{cl} is given by

<<>>=
n <- 10
rep(1, 10)
@



\section{Significance Analysis of Microarrays}

In this section, we show how the Significance Analysis of
Microarrays (SAM) proposed by Tusher et al.\ (2001) can be applied
to a data set. We start with the version of SAM that uses a
modified version of the usual $t$ statistic. This modification is
necessary to avoid that genes with low expression levels dominate
the results of the SAM analysis. Since the SAM analysis using
Wilcoxon rank sums described in Schwender (2003) works in a
similar way, we only take a short look on this analysis.

\subsection{SAM Analysis with \texttt{sam}, \texttt{sam.delta} and \texttt{sam.plot}}

As mentioned in the introduction, the Golub et al.\ (1999) data
set provided by the \texttt{multtest} package is used as our example
data set. The matrix \texttt{golub} contains the expression values of the
3051 genes and the 38 samples, and the vector \texttt{golub.cl}
consists of the class labels of the samples. Additionally, we use
the mean instead of the median number of falsely called genes over
the \texttt{B} permutations (here the default \texttt{B=100} is
used) in the estimation of the number of false positives (for
details, see Schwender 2003), and hence set \texttt{med=FALSE}.
Specifying \texttt{rand} sets the random number generator in a
reproducible state. For further analyses, it is necessary to assign
the output of \texttt{sam} to an object (here \texttt{sam.out}).
Using these settings, the SAM analysis is performed by

<<>>=
sam.out <- sam(golub, golub.cl, med = FALSE, rand = 123)
@

\newpage

\begin{figure}[!h]
\small \renewcommand{\baselinestretch}{1.2}
\centerline{
\includegraphics[width=0.6\textwidth]{samplot1}}\vspace{-12pt}
\caption{SAM Plot. Plot of the observed ordered expression scores
$d_i$ vs.\ the expected ordered expression scores $\bar{d}_i$.}
\end{figure}
\vspace*{48pt}

\begin{figure}[!h]
\small \renewcommand{\baselinestretch}{1.2} \centerline{
\includegraphics[width=0.8\textwidth,height=0.5\textwidth]{delta}} \vspace{-12pt}
\caption{Delta Plots. These two plots visualize the table in the
output of \texttt{sam}.}
\end{figure}

\newpage

The output of \texttt{sam} consists of the two plots displayed in
Figure 1 and Figure 2, and a table containing the number of significant
genes, \texttt{called}, the estimated \texttt{FDR}, the estimated
value of the prior probability \texttt{p0} that a gene is
not differentially expressed, and the mean (or median) number,
\texttt{false}, of falsely called genes for several values of the
threshold $\Delta$. The number of false positives can be computed
by multiplying \texttt{false} with \texttt{p0}. This output also
contains the value of the fudge factor $s_0$ which is the constant that is
added to the denominator of the usual $t$ statistic.

In the original setting, Tusher et al.\ (2001) define the fudge
factor as the quantile of the standard deviations of the genes that
is in some sense optimal (for the optimization criterion, see Tusher et al.\
2001), and $s_0$ is thus a small \textit{strictly} positive constant. As
the output of \texttt{sam} shows, we here also allow the fudge factor
to become zero. The choice $s_0=0$ can be avoided, and hence the
original definition of Tusher et al.\ (2001) can be used by
setting \texttt{include.s0=FALSE}.

Either the table or the Delta plots in Figure 2 which are a
visualization of the table can be used to choose the value of
$\Delta$ that provides the best balance between the number of
identified genes and the estimated FDR. Since this selection
depends on the goal of the user's analysis, we cannot give a
recommendation on how the optimal value of $\Delta$ should be
chosen.

If one would like to choose a value of $\Delta$ that is more specific than
the values provided by the default setting of \texttt{sam}, i.e.\
\texttt{delta=(1:10)/5}, one can use the function
\texttt{sam.delta} to obtain the number of genes called differentially
expressed and the estimated FDR for other choices of $\Delta$.
Assume, e.g., that the goal of our analysis is to call all genes
 differentially expressed that have an estimated FDR less or equal to 0.01. In this case, we
can select 1.8 as our value for $\Delta$ which leads to 438 genes
called significant and an estimated FDR of 0.007, or we can use
\texttt{sam.delta} for selecting a value of $\Delta$ that results
in an FDR that is closer to 0.01. Such an analysis can be
performed by

<<>>=
sam.delta(sam.out, seq(1.6, 1.8, 0.02))
@



\vspace*{18pt} \noindent where the required arguments are the
object to which the output of \texttt{sam} has been assigned and a
set of values for the threshold $\Delta$.

Let's say our choice of $\Delta$ is 1.70. This choice leads to 513
genes called differentially expressed. We expect that 0.49*10.42
of these genes are false positives, i.e.\ that 5.1058 of these
genes are not really differentially expressed. The estimated value of
FDR is thus 0.010. Now the function \texttt{sam.plot} can be used to
obtain gene-specific statistics as the expression scores $d$, the
$q$-values and the R-folds of the genes called
significant\footnote{Warning: Using this command interactively
will save the output to the file "sam.txt" in your local
directory.}.

<<>>=
sam.plot.out <- sam.plot(sam.out, 1.7,
gene.names = golub.gnames[,2], file = "sam.txt")
@


The required arguments of \texttt{sam.plot} are the object to
which the output of \texttt{sam} has been assigned and a value
of $\Delta$, where this value of $\Delta$ need not necessarily
have been used in the analysis with \texttt{sam} (or \texttt{sam.delta}).
The names of the differentially expressed genes are added to the
table of the gene-specific statistics by setting
\texttt{gene.names=golub.gnames[,2]}. This table and additional
general statistics as the number of differentially expressed genes
and the estimated FDR will be stored in a file, if \texttt{file}
is specified. Here this output is stored in the file
\texttt{sam.txt}.

\begin{figure}[!h]
\vspace*{24pt} \small \renewcommand{\baselinestretch}{1.2}
\centerline{
\includegraphics[width=0.55\textwidth]{samplot2}
} \vspace{-12pt} \caption{SAM plot for $\Delta=1.7$. Significant
genes are marked green. }\label{samplot2} \vspace*{42pt}
\end{figure}


\begin{figure}[!h]
\small \renewcommand{\baselinestretch}{1.2}
 \centerline{
\includegraphics[width=0.9\textwidth]{samplotout}}
\vspace{-4pt}\caption{General information on the differentially
expressed genes and gene-specific statistics for the
differentially expressed genes stored in a
file.}\label{samplotout} %\vspace*{36pt}
\end{figure}


Besides from storing important information to a file (an excerpt
of this file is displayed in Figure \ref{samplotout}),
\texttt{sam.plot} generates the SAM plot for the specified choice
of $\Delta$ (see Figure \ref{samplot2}). For a comparison with
other methods or for further analyses, it is necessary to assign
the output of \texttt{sam.plot} to an object. This object can be
used to display the table that has been stored in \texttt{file} in
the \texttt{R} console,\\[18pt]
\texttt{>sam.plot.out\$sam.output}\\[18pt]
or to obtain the general information,\\[18pt]
\texttt{> sam.plot.out\$vec.fdr}\\[18pt]
or to obtain the vector \texttt{row.sig.genes} that consists of
the rows of the data set that contain the expression data of the
differentially expressed genes.\\[18pt]
\texttt{> sam.plot.out\$row.sig.genes}




\subsection{SAM Analysis with \texttt{sam.wilc} and \texttt{sam.plot}}

Instead of a (modified) $t$ statistic, the Wilcoxon rank sum
statistic can be used for the identification of genes that are
differentially expressed between two groups. In Schwender (2003),
it is described how the original SAM algorithm can be modified for
Wilcoxon rank sums.

Such an SAM analysis is implemented in the function
\texttt{sam.wilc} whose usage is similar to the usage of
\texttt{sam}. A SAM analysis using Wilcoxon rank sums is
thus performed by\\[18pt]
\texttt{>sam.wilc.out<-sam.wilc(golub,golub.cl,rand=123)}\\[18pt]
Since \texttt{sam.wilc} by default computes the number of
differentially expressed genes and the estimated FDR for all
possible values of $\Delta$, no further analysis with
\texttt{sam.delta} is required, and the optimal value of $\Delta$
can chosen by taking a look on the table or the Delta plots
provided by \texttt{sam.wilc}. Using this value, \texttt{sam.plot}
can be used to obtain specific information on the differentially
expressed genes\footnote{Warning: Using this command interactively
will save the output to the file "sw.txt" in your local
directory.}.
<<echo=F,results=hide>>=
sam.wilc.out<-sam.wilc(golub,golub.cl,rand=123)
@

<<>>=
sam.wilc.plot.out <- sam.plot(sam.wilc.out, 48, file = "sw.txt",
gene.names = golub.gnames[, 2])
@
\vspace{18pt}

\section{Empirical Bayes Analysis of Microarrays}

The original version of the Empirical Bayes Analysis of
Microarrays proposed by Efron et al.\ (2001) is based on the same
modified $t$-statistic also used in SAM. In another paper,
Efron et al.\ (2002) modify this version of EBAM by replacing
the modified $t$ statistic by the Wilcoxon rank sum statistic. In
this section, it is shown how these two versions of EBAM can be
applied to a data set.

\subsection{EBAM with \texttt{find.a0} and \texttt{ebam}}

For their EBAM analysis, Efron et al.\ (2001) summarize the
expression values of each gene by the same modified version of the
usual $t$ statistic that is computed in SAM. The only difference
lies in the computation of the fudge factor. While in the SAM
procedure this computation is automatically done by \texttt{sam},
we here need to specify the fudge factor prior to performing the
main EBAM analysis. Actually, computing the fudge factor $a_0$ for
the EBAM analysis means to perform a (standardized) EBAM analysis
for each specified value of $a_0$, and then selecting the value
that is in some sense optimal. For such a comparison, it is
necessary to always have the same marginal distribution for the
permuted expression scores. For each value of $a_0$, both the
observed and the permuted expression scores are therefore
transformed such that the permuted expression scores follow a
standard normal distribution. This analysis can be performed by

\begin{verbatim}
> find.out <- find.a0(golub, golub.cl, rand = 123)

EBAM Analysis for the two class unpaired case.


 Number of significant genes for some a0:
                 a0=0   a0=0.0606 (alpha=0) a0=0.1148 (alpha=0.1)
                  740                   740                   680
a0=0.1292 (alpha=0.2) a0=0.1428 (alpha=0.3) a0=0.1562 (alpha=0.4)
                  680                   680                   680
a0=0.1695 (alpha=0.5) a0=0.1849 (alpha=0.6) a0=0.2025 (alpha=0.7)
                  649                   649                   649
a0=0.2272 (alpha=0.8) a0=0.2707 (alpha=0.9)
                  623                   623

 Suggested choice for a0: 0
\end{verbatim}

\begin{figure}[!h]
\vspace{18pt} \small \renewcommand{\baselinestretch}{1.2}
\centerline{
\includegraphics[width=.6\textwidth]{logitpost}}\vspace{-12pt}
\caption{Plot of the transformed (observed) expression scores vs.\
the logit of their posterior probabilities.}\label{logitpost}
%\vspace*{18pt}
\end{figure}

The output of \texttt{find.a0} suggests a value for
$a_0$ (here $a_0=0$). This suggestion follows the optimization
criterion of Efron et al. (2001) who argue that the value of $a_0$
should be selected that leads to the most differentially expressed
genes. If there are more than one optimal choice, \texttt{find.a0}
will suggest the smallest of these values.

One, however, should also take a look on the plot of the
transformed observed expression scores vs.\ the logit of their
posterior probabilities (see Figure \ref{logitpost}). If another
value than the suggested value leads to a slightly smaller number
of differentially expressed genes, but has higher posterior
probabilities for the most extreme expression values, then it will
also be appropriate to use this value, since the latter indicates
a better separation between the distribution of all genes and the
distribution of the not differentially expressed genes (cf.\ Efron
et al.\ 2001).



For our analysis with \texttt{ebam}, we use the suggested choice
for $a_0$. Since this value is used as default in
\texttt{ebam}, we do not have to specify \texttt{a0}. Other choices
must be specified. The usage
and the output of \texttt{ebam} is similar to the usage and the
output of \texttt{sam.plot}.

\begin{verbatim}
> ebam.out <- ebam(find.out, gene.names = golub.gnames[, 2],
+ file="sam.txt")
 Using a0 = 0 and the original Z values, there are
 714 significant genes and 37.75 falsely called genes.
 For p0 = 0.4901 , the FDR is 0.0259 .

Output is stored in ebam.txt
\end{verbatim}

\begin{figure}[!ht]
\small \renewcommand{\baselinestretch}{1.2}
\centerline{
\includegraphics[width=0.5\textwidth]{ebam}}\vspace{-12pt}
\caption{Plot of the expression scores vs.\ their posterior
probability. Genes called differentially expressed are marked
green.}\label{ebam}\vspace{18pt}
\end{figure}


For each differentially expressed gene, its
expression value $z$, its posterior probability, its $q$-value,
its $R$-fold, and two estimates of its local FDRs (cf.\ Efron et
al. 2001) are stored in the output file. Furthermore,
\texttt{ebam} generates the plot of the posterior probabilities of
the genes in which the differentially expressed genes are marked
green (see Figure \ref{ebam}). The number of genes called differentially
expressed may differ between \texttt{find.a0} and \texttt{ebam}
since in the former the transformed and in the latter the original
expression scores are used. The rows of the data matrix containing
the differentially expressed genes can -- similar to
\texttt{sam.plot} -- be obtained by \texttt{ebam.out\$row.sig.genes}.

\subsection{EBAM with \texttt{ebam.wilc}}

Contrary to the other three analyses, only one function, namely
\texttt{ebam.wilc}, has to be called for the EBAM analysis using
Wilcoxon rank sums (for details of this procedure, see Efron et
al.\ 2002)\footnote{Warning: Using this command interactively will
save the output to the file "ebam.wilc.txt" in your local
directory.}.

<<>>=
ebam.wilc.out <- ebam.wilc(golub, golub.cl, g = golub.gnames[,2],
rand = 123, file = "ebam.wilc.txt")
@

Again gene-specific statistics of the differentially expressed
genes and general information on, e.g., the number of
differentially expressed genes and the estimated FDR will be
stored in a file, if \texttt{file} is specified. The general
information are also displayed by \texttt{ebam.wilc} in the
\texttt{R} console. Furthermore, \texttt{ebam.wilc} displays the
number of genes having a tied Wilcoxon rank sum. The expression
score of each of these genes is by default randomly assigned
either to the next larger or smaller integer. Not displayed are
the two plots generated by \texttt{ebam.wilc}. This is, on the one
hand, the plot of the null and the mixture density (both
multiplied by the number of genes) of the expression scores, and
on the other hand, the plot of the posterior probabilities of the
genes in which the differentially expressed genes are marked
green.

%\vspace{1cm}

\begin{thebibliography}{}
\item[Efron, B.,] Tibshirani, R., Storey, J.\ D., and Tusher, V.\ (2001),
``Empirical Bayes Analysis of a Microarray Experiment,"
\textit{Journal of the American Statistical Association}, 96,
1151--1160.

\item[Efron, B.,] and Tibshirani, R.\ (2002), ``Empirical Bayes Methods
and False Discovery Rates for Microarrays," \textit{Genetic
Epidemiology}, 23, 70--86.

\item[Golub, T.\ R.,] Slonim, D.\ K., Tamayo, P., Huard, C., Gaasenbeek,
M., Mesirov, J.\ P., Coller, H., Loh, M.\ L., Downing, J.\ R.,
Caliguiri, M.\ A., Bloomfield, C.\ D., and Lander, E.\ S. (1999),
``Molecular Classification of Cancer: Class Discovery and Class
Prediction by Gene Expression Moni\-to\-ring," \textit{Science},
286, 531--537.

\item[Schwender, H.] (2003), ''Assessing the False Discovery Rate
in a Statistical Analysis of Gene Expression Data,"
\textit{Diploma thesis}, University of Dortmund,
\texttt{http://de.geocities.com/holgerschw/thesis.pdf}.

\item[Tusher, V.\ G.,] Tibshirani, R., and Chu, G.\ (2001),
``Significance Analysis of Microarrays Applied to the Ionizing
Radiation Response," \textit{Proceedings of the National Academy
of Science,} 98, 5116--5121.

\end{thebibliography}



\end{document}
